extensions/CoreWidgetProvider/Helpers/GPUStats.cs

```diff
@@ -99,23 +101,39 @@ public void GetData()
             List<PerformanceCounter>? counters;
             var success = _gpuCounters.TryGetValue(gpu.PhysId, out counters);
 
-            if (success)
+            if (success && counters != null)
             {
-                try
+                var sum = 0.0f;
+                var countersToRemove = new List<PerformanceCounter>();
+                foreach (var counter in counters)
                 {
-                    // NextValue() can throw an InvalidOperationException if the counter is no longer there.
-                    var sum = counters?.Sum(x => x.NextValue()) ?? 0;
-                    gpu.Usage = sum / 100;
-                    lock (gpu.GpuChartValues)
+                    try
+                    {
+                        // NextValue() can throw an InvalidOperationException if the counter is no longer there.
+                        sum += counter.NextValue();
+                    }
+                    catch (InvalidOperationException ex)
+                    {
+                        // We can't modify the list during the loop, so save it to remove at the end.
+                        _log.Information(ex, "Failed to get next value, remove");
+                        countersToRemove.Add(counter);
+                    }
+                    catch (Exception ex)
                     {
-                        ChartHelper.AddNextChartValue(sum, gpu.GpuChartValues);
+                        _log.Error(ex, "Error going through process counters.");
                     }
                 }
-                catch (InvalidOperationException ex)
+
+                foreach (var counter in countersToRemove)
+                {
+                    counters.Remove(counter);
+                    counter.Dispose();
+                }
+
+                gpu.Usage = sum / 100;
+                lock (gpu.GpuChartValues)

```

Comment:
Assuming we're locking here because `GetData()` can be called from multiple threads, right? In this case, should we also lock on other shared resources as well such as the value obtain from `_gpuCounters`?

Reply:
I think this would make sense as part of a larger overhaul, I think there's a lot we want to sort out with these widgets.

Reply:
- Do we have an issue tracking this change? If not, please create one.
- Since the `foreach` is outside the `try/catch` and there is no `lock`, this can potentially cause an exception/crash.

Reply:
Spoke offline, put back an outer try/catch for now.